modAionImpl/src/org/aion/zero/impl/sync/SyncStats.java

```diff
@@ -23,27 +23,268 @@
 
 package org.aion.zero.impl.sync;
 
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
 /** @author chris */
-final class SyncStats {
+public final class SyncStats {
 
     private long start;
 
     private long startBlock;
 
     private double avgBlocksPerSec;
 
+    private final ConcurrentHashMap<String, Long> requestsToPeers = new ConcurrentHashMap<>();
+
+    private final ConcurrentHashMap<String, Long> blocksByPeer = new ConcurrentHashMap<>();
+
+    private final ConcurrentHashMap<String, Long> blockRequestsByPeer = new ConcurrentHashMap<>();
+
+    private final ConcurrentHashMap<String, LinkedList> statusRequestTimeByPeers =
+            new ConcurrentHashMap<>();
+
+    private final ConcurrentHashMap<String, LinkedList> statusResponseTimeByPeers =
+            new ConcurrentHashMap<>();
+
+    private Long overallAvgPeerResponseTime;
+
     SyncStats(long _startBlock) {
         this.start = System.currentTimeMillis();
         this.startBlock = _startBlock;
         this.avgBlocksPerSec = 0;
+        this.overallAvgPeerResponseTime = 0L;
     }
 
-    synchronized void update(long _blockNumber) {
+    /**
+     * Update statistics based on peer nodeId, total imported blocks, and best block number
+     *
+     * @param _nodeId peer node display Id
+     * @param _totalBlocks total imported blocks in batch
+     * @param _blockNumber best block number
+     */
+    synchronized void update(String _nodeId, int _totalBlocks, long _blockNumber) {
         avgBlocksPerSec =
                 (double) (_blockNumber - startBlock) * 1000 / (System.currentTimeMillis() - start);
+        updateTotalRequestsToPeer(_nodeId);
+        updatePeerTotalBlocks(_nodeId, _totalBlocks);
     }
 
     synchronized double getAvgBlocksPerSec() {
         return this.avgBlocksPerSec;
     }
+
+    /**
+     * Updates the total requests made to a pear
+     *
+     * @param _nodeId peer node display Id
+     */
+    private void updateTotalRequestsToPeer(String _nodeId) {
+        if (requestsToPeers.putIfAbsent(_nodeId, 1L) != null) {
+            requestsToPeers.computeIfPresent(_nodeId, (key, value) -> value + 1L);
+        }
+    }
+
+    /**
+     * Calculates the percentage of requests made to each peer with respect to the total number of
+     * requests made.
+     *
+     * @return a hash map in descending order containing peers with underlying percentage of
+     *     requests made by the node
+     */
+    synchronized Map<String, Float> getPercentageOfRequestsToPeers() {
```

Comment:
While testing the execution on the mastery test network I got the following exception:
```
18-11-05 17:30:46.450 INFO  P2P  [sync-ss]: sync-status avg-import=32.68 b/s td=31115/55526982 b-num=1923/894870 b-hash=042fe74a77fa53c72fc5c629c985eb5588856a1ed6a8a437f8a02a9655eb36fa/268c9c28ea547c14c4eca6efaaa795f823700017b51c956ad38753dd98cf438e
Exception in thread "sync-ss" java.lang.ArrayIndexOutOfBoundsException: 12
	at java.base/java.util.stream.SortedOps$SizedRefSortingSink.accept(SortedOps.java:369)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1751)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.aion.zero.impl.sync.SyncStats.getPercentageOfRequestsToPeers(SyncStats.java:115)
	at org.aion.zero.impl.sync.TaskShowStatus.dumpPeerStatsInfo(TaskShowStatus.java:161)
	at org.aion.zero.impl.sync.TaskShowStatus.run(TaskShowStatus.java:120)
	at java.base/java.lang.Thread.run(Thread.java:844)
```